### PR TITLE
GPII-2833 - Add persistence to auto generated variables

### PR DESCRIPTION
**Pull request by amatas**
This PR includes some refactoring of the code and also saves the variables in a yml file if they aren't found. In the case that the variables are set using environment variables (i.e. CI production or staging) the variables are not saved in a local file to improve the security.

More info: https://issues.gpii.net/browse/GPII-2833

### CODE REVIEW

**Comment by mrtyler** [comment]
Thanks for the quick turnaround Alfredo! This looks good.

However, I think a lot of things could be simplified by using strings, rather than symbols, as keys in `@secrets`. In particular, I think a lot of massaging could be replaced with `YAML.dump` or `.to_yaml`.

Is there an advantage to using symbols that makes the extra complexity worthwhile?

**Comment by amatas** [comment]
There isn't a particular reason to use symbols. I wrote this code in order to have a quick patch with the minor impact in all the files due to the rush. But, of course, we can get rid of the symbols and save some lines of code.

**Comment by mrtyler** [comment]
I hear you, and I appreciate the effort.

I think it makes sense to improve this fix while it's fresh in your mind, and before it confuses a gpii-infra developer or user. Thanks!

**Comment by mrtyler** [comment]
BTW, I tried your fix locally to spin up a dev cluster and it worked -- generated passwords were persisted so the cluster did not break itself after a subsequent `rake`.

**Comment by amatas** [comment]
> BTW, I tried your fix locally to spin up a dev cluster and it worked -- generated passwords were persisted so the cluster did not break itself after a subsequent rake.

Awesome, I'm working on the evolution of this PR.

**Comment by amatas** [comment]
The last commit is a proposal about how to manage secrets using a "Secrets" object

File: modules/deploy/52-couchdb-secrets.erb [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+kind: Secret
+metadata:
+  namespace: gpii
+  name: couchdb-secrets
+type: Opaque
+data:
+  user: <%= Base64.strict_encode64(@secrets["couchdb_user"]) %>
+  password: <%= Base64.strict_encode64(@secrets["couchdb_password"]) %>
+  erlang_cookie: <%= Base64.strict_encode64(@secrets["erlang_cookie"]) %>
+  url-preferences: <%= Base64.strict_encode64("http://#{@secrets["couchdb_user"]}:#{@secrets["couchdb_password"]}@couchdb-lb.gpii.svc.cluster.local:5984/preferences") %>
```
Comment (mrtyler) [inline modules/deploy/52-couchdb-secrets.erb]: couchdb_preferences_url

File: modules/deploy/52-couchdb-secrets.erb [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+  url-preferences-datasource: <%= Base64.strict_encode64("http://#{@secrets["couchdb_user"]}:#{@secrets["couchdb_password"]}@couchdb-lb.gpii.svc.cluster.local:5984/preferences/%userToken") %>
```
Comment (mrtyler) [inline modules/deploy/52-couchdb-secrets.erb]: preferences_datasource_url

**Comment by ussjoin** [comment]
Looks fine to me; obviously this is just an intermediate step in moving secret data to somewhere more safe, but I appreciate Alfredo's work in doing the refactoring now to make that next jump easier.

**Pull request merged**